lib/puppet_x/templates/powershell/init_ps.ps1

```diff
@@ -748,6 +748,8 @@ function Start-PipeServer
     $Encoding
   )
 
+  Add-Type -AssemblyName "System.Core, Version=3.5.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"
```

Comment:
What about for systems that don't have the `3.5.0.0` version of `System.Core`?

Reply:
I suspect PowerShell has binding redirects setup, but I have no confirmation of this yet ... On our 2016 templates, there is no `System.Core 3.5.0.0` - but `System.Core 4.0.30319` is already loaded into the PowerShell process AppDomain. The call to `Add-Type` with the explicit version does not fail. However, its unclear if that's due to `System.Core` already being loaded / the call being ignored or if its due to first redirecting the `3.5.0.0` request to a `4.0.30319` request and then determining the assembly is already loaded.

More info on binding redirects:
https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/redirect-assembly-versions

I'm digging in with the fusion log viewer to see if I can get any more details:
https://docs.microsoft.com/en-us/dotnet/framework/tools/fuslogvw-exe-assembly-binding-log-viewer

Reply:
As a quick follow-up, I've confirmed behavior using the fusion log viewer. When requesting `System.Core 3.5.0.0`, fusion redirects that request to `System.Core, Version=4.0.0.0`. Since that assembly is already loaded, `Add-Type` succeeds:

```
*** Assembly Binder Log Entry  (3/27/2018 @ 3:35:39 PM) ***

The operation was successful.
Bind result: hr = 0x0. The operation completed successfully.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Core, Version=3.5.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
 (Fully-specified)
LOG: Appbase = file:///C:/Windows/System32/WindowsPowerShell/v1.0/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = powershell.exe
Calling assembly : (Unknown).
===
LOG: This bind starts in default load context.
LOG: No application configuration file found.
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Version redirect found in framework config: 3.5.0.0 redirected to 4.0.0.0.
LOG: Post-policy reference: System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
LOG: Binding succeeds. Returns assembly from C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Core\v4.0_4.0.0.0__b77a5c561934e089\System.Core.dll.
LOG: Assembly is loaded in default load context.
```

I suspect that without a redirect in place the load will instead fail, so I think the best bet here is probably to stick with the simpler `Add-Type -AssemblyName 'System.Core'` so that for machines that may not perform redirects back to 3.5.0.0, the code will still work (which might be the case in a PS6 or cross-platform scenario). We're less concerned with the version of the assembly, and more concerned with gaining access to a few types within it.